src/dbSta/src/dbNetwork.cc

```diff
@@ -687,71 +749,161 @@ InstanceNetIterator* dbNetwork::netIterator(const Instance* instance) const
 
 ObjectId dbNetwork::id(const Pin* pin) const
 {
-  dbITerm* iterm;
-  dbBTerm* bterm;
-  staToDb(pin, iterm, bterm);
+  dbITerm* iterm = nullptr;
+  dbBTerm* bterm = nullptr;
+  dbModITerm* moditerm = nullptr;
+  dbModBTerm* modbterm = nullptr;
+
+  staToDb(pin, iterm, bterm, moditerm, modbterm);
 
   if (iterm != nullptr) {
     return iterm->getId() << 1;
   }
-  return (bterm->getId() << 1) + 1;
+  if (bterm != nullptr) {
+    return (bterm->getId() << 1) + 1;
+  }
+
+  if (moditerm != nullptr) {
+    return (moditerm->getId());
+  }
+  if (modbterm != nullptr) {
+    return (modbterm->getId());
+  }
+  return 0;
 }
 
 Instance* dbNetwork::instance(const Pin* pin) const
 {
   dbITerm* iterm;
   dbBTerm* bterm;
-  staToDb(pin, iterm, bterm);
+  dbModITerm* moditerm = nullptr;
+  dbModBTerm* modbterm = nullptr;
+
+  staToDb(pin, iterm, bterm, moditerm, modbterm);
   if (iterm) {
     dbInst* dinst = iterm->getInst();
     return dbToSta(dinst);
   }
   if (bterm) {
     return top_instance_;
   }
+  if (moditerm) {
+    dbModInst* mod_inst = moditerm->getParent();
+    return dbToSta(mod_inst);
+  }
+  if (modbterm) {
+    dbModule* module = modbterm->getParent();
+    dbModInst* mod_inst = module->getModInst();
+    return dbToSta(mod_inst);
+  }
   return nullptr;
 }
 
 Net* dbNetwork::net(const Pin* pin) const
 {
-  dbITerm* iterm;
-  dbBTerm* bterm;
-  staToDb(pin, iterm, bterm);
+  dbITerm* iterm = nullptr;
+  dbBTerm* bterm = nullptr;
+  dbModITerm* moditerm = nullptr;
+  dbModBTerm* modbterm = nullptr;
+
+  staToDb(pin, iterm, bterm, moditerm, modbterm);
   if (iterm) {
     dbNet* dnet = iterm->getNet();
-    return dbToSta(dnet);
+    dbModNet* mnet = iterm->getModNet();
+    // It is possible when writing out a hierarchical network
+    // that we have both a mod net and a dbinst net.
+    // In the case of writing out a hierachical network we always
+    // choose the mnet.
+    if (dnet && mnet) {
+      return dbToSta(mnet);
+    }
+    if (mnet)
+      return dbToSta(mnet);
+    if (dnet)
+      return dbToSta(dnet);
+  }
+  // only pins which act as bterms are top levels and have no net
+  if (bterm) {
+    return nullptr;
+  }
+  if (moditerm) {
+    dbModNet* mnet = moditerm->getModNet();
+    return dbToSta(mnet);
+  }
+  if (modbterm) {
+    dbModNet* mnet = modbterm->getModNet();
+    return dbToSta(mnet);
   }
   return nullptr;
 }
 
 Term* dbNetwork::term(const Pin* pin) const
 {
-  dbITerm* iterm;
-  dbBTerm* bterm;
-  staToDb(pin, iterm, bterm);
+  dbITerm* iterm = nullptr;
+  dbBTerm* bterm = nullptr;
+  dbModITerm* moditerm = nullptr;
+  dbModBTerm* modbterm = nullptr;
+  staToDb(pin, iterm, bterm, moditerm, modbterm);
   if (iterm) {
     return nullptr;
   }
   if (bterm) {
     return dbToStaTerm(bterm);
   }
+  if (moditerm) {
+    // get the mod bterm
+    std::string port_name_str = moditerm->getName();
+    size_t last_idx = port_name_str.find_last_of('/');
```

Comment:
It would be better to use parallel arrays in the future.

Reply:
and if the modbterm and moditerm ids don't use the weird shift odd/even thing (see above) that is possible. I wonder if we can avoid the shift thing for the ids and assume that the modbterms and moditerms are created in linear order using the tight db arrays (linked list order same).